File: composeApp/src/androidMain/kotlin/com/snapchef/app/features/profile/presentation/ProfileViewModel.kt
package com.snapchef.app.features.profile.presentation

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONArray
import org.json.JSONObject

data class ProfileSavedRecipe(
    val title: String,
    val isQuick: Boolean,
)

data class ProfileUiState(
    val recipes: List<ProfileSavedRecipe> = listOf(
        ProfileSavedRecipe("Omelette with Cheese", true),
        ProfileSavedRecipe("Tomato Egg Fried Rice", true),
        ProfileSavedRecipe("Baked Veggie Pasta", false),
        ProfileSavedRecipe("Leftover Chicken Wraps", true),
    ),
)

class ProfileViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(ProfileUiState())
    val uiState: StateFlow<ProfileUiState> = _uiState.asStateFlow()

    fun applyBackendJson(json: String) {
        runCatching {
            val obj = JSONObject(json)
            val array = obj.optJSONArray("recipes") ?: JSONArray()
            val mapped = buildList {
                for (i in 0 until array.length()) {
                    val item = array.optJSONObject(i) ?: continue
                    add(
                        ProfileSavedRecipe(
                            title = item.optString("title", "Untitled recipe"),
                            isQuick = item.optBoolean("isQuick", false),
                        ),
                    )
                }
            }
            if (mapped.isNotEmpty()) _uiState.value = _uiState.value.copy(recipes = mapped)
        }
    }
}

File: iosApp/iosApp/ui/ProfileScreen/ProfileViewModel.swift
//
//  ProfileViewModel.swift
//  iosApp
//
//  Created by gergana on 3/27/26.
//

import Foundation
import Combine

struct ProfileSavedRecipe: Identifiable {
    let id = UUID()
    let title: String
    let isQuick: Bool
}

struct ProfileUiState {
    var recipes: [ProfileSavedRecipe] = [
        ProfileSavedRecipe(title: "Omelette with Cheese", isQuick: true),
        ProfileSavedRecipe(title: "Tomato Egg Fried Rice", isQuick: true),
        ProfileSavedRecipe(title: "Baked Veggie Pasta", isQuick: false),
        ProfileSavedRecipe(title: "Leftover Chicken Wraps", isQuick: true)
    ]
}

@MainActor
class ProfileViewModel: ObservableObject {
    @Published var uiState = ProfileUiState()
    
    func applyBackendJson(json: String) {
        guard let data = json.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let array = dict["recipes"] as? [[String: Any]] else {
            return
        }
        
        let mapped = array.compactMap { item -> ProfileSavedRecipe? in
            let title = item["title"] as? String ?? "Untitled recipe"
            let isQuick = item["isQuick"] as? Bool ?? false
            return ProfileSavedRecipe(title: title, isQuick: isQuick)
        }
        
        if !mapped.isEmpty {
            uiState.recipes = mapped
        }
    }
}
